contracts/modules/TransferManager/VolumeRestrictionTM.sol

```diff
@@ -0,0 +1,934 @@
+        // Check with the bucket and parse all the new timestamps to calculate the sumOfLastPeriod
+        // re-using the local variables to avoid the stack too deep error.
+        (sumOfLastPeriod, fromTimestamp, passedTimestamps) = _bucketCheck(
+            fromTimestamp,
+            diffDays,
+            _from,
+            globalBucketDetails,
+            true,
+            globalRestriction
+        );
+        // Total amout that is transacted uptill now for `fromTimestamp` day
+        uint256 txSumOfDay = globalBucket[fromTimestamp];
+        // Global transaction check
+        if (_globalTxCheck(sumOfLastPeriod, txSumOfDay, _amount)) {
+            // allow modification in storage when `_isTransfer` equals true
+            if(_isTransfer) {
+                // update the global storage
+                _updateGlobalStorage(
+                    passedTimestamps,
+                    fromTimestamp,
+                    _amount,
+                    diffDays,
+                    sumOfLastPeriod
+                );
+            }
+            return Result.NA;
+        }
+        return Result.INVALID;
+    }
+
+    /**
+     * @notice Internal function to update the state variables related to global restriction
+     */
+    function _updateGlobalStorage(
+        uint256[] passedTimestamps,
+        uint256 _fromTime,
+        uint256 _amount,
+        uint256 _diffDays,
+        uint256 _sumOfLastPeriod
+    )
+        internal
+    {
+        if (_diffDays != 0) {
+            for (uint256 i = 0; i < passedTimestamps.length; i++) {
+                // Add the timestamp that is already passed
+                globalBucketDetails.timestamps.push(passedTimestamps[i]);
+            }
+        }
+        // This condition is the works only when the transaction performed just after the startTime (_diffDays == 0)
+        if (globalBucketDetails.timestamps.length == 0) {
+            globalBucketDetails.timestamps.push(_fromTime);
+        }
+        if(_amount != 0) {
+            // updating the sumOfLastPeriod
+            globalBucketDetails.sumOfLastPeriod = _sumOfLastPeriod.add(_amount);
+
+            // Re-using the local variable to avoid stack too deep error
+            _fromTime = globalBucketDetails.timestamps[globalBucketDetails.timestamps.length -1];
+            globalBucket[_fromTime] = globalBucket[_fromTime].add(_amount);
+        }
+    }
+
+    /// Internal function for the bucket check
+    function _bucketCheck(
+        uint256 _fromTime,
+        uint256 _diffDays,
+        address _from,
+        BucketDetails _bucketDetails,
+        bool _isGlobal,
+        VolumeRestriction _restriction
+    )
+        internal
+        view
+        returns (uint256, uint256, uint256[])
+    {
+        uint256[] memory passedTimestamps = new uint256[](_diffDays);
+        uint256 counter = _bucketDetails.timestamps.length;
+        uint256 i = 0;
+        if (_diffDays != 0) {
```

Comment:
This check is not required as it is already being checked on next line in the for loop.